Add CLI tool for PR review optimization

minaliuTT opened the pull request:
### Problem description
This PR introduces a CLI tool to optimize PR reviewers by automatically selecting reviewers based on code ownership. The tool:
1. Retrieves the list of files changed in a PR.
2. Identifies code owners for each file.
3. Allows including or skipping specific reviewers.
4. Selects an optimal set of reviewers based on file coverage.

### What's changed
✅ Added a new CLI tool under `cli_tool/`
✅ Implemented `review_optimizer.py` to handle reviewer selection
✅ Added `setup.py` for packaging
✅ Included `README.md` with usage instructions

### For Development (Editable Install)

```bash
cd /path/to/tt-metal/cli_tool
pip install -e .
```
### For Regular Use

```bash
cd /path/to/tt-metal/cli_tool
pip install .
```
### Usage

After installation, you can use the tool from anywhere with the following command syntax:
```bash
get_reviewers PR_NUMBER -include name1 name2 ... -skip name3 name4 ...
```

```bash
get_reviewers PR_NUMBER
```

Where:
- `PR_NUMBER` is the number of the pull request. **(Mandatory)**
- `-include` allows you to specify reviewers to include in the selection process. **(Optional)**
- `-skip` allows you to specify reviewers to skip from the selection process. **(Optional)**

### Limitations
This initial version does **not** dereference group members. Only **individual** usernames can be included or skipped in the reviewer selection.

Commit:
Add CLI tool for PR review optimization

blozano-tt commented:
1. This could be a very useful PR
2. Do you have an example of what it does? Screen shot? What does select a reviewer mean?
3. We should avoid adding yet another top level folder to this repo. Can this tool live somewhere else?

cc: @ayerofieiev-tt @bkeith-TT

tt-rkim commented on cli_tool/pr_review/review_optimizer.py:
```diff
@@ -0,0 +1,111 @@
+    """Get codeowners for a file path"""
+    try:
+        result = subprocess.run(
+            ["codeowners", file_path],
```
Is this a command? How do people get access to it? Is it a package?

minaliuTT replied:
This is a third-party [tool](https://github.com/hmarr/codeowners) for parsing CODEOWNERS files. Can we use it?

tt-rkim commented on cli_tool/setup.py:
```diff
@@ -0,0 +1,22 @@
+            'get_reviewers=pr_review.review_optimizer:review_optimizer',
+        ],
+    },
+)
```
posix newline, pls delete

tt-rkim commented on cli_tool/pr_review/__init__.py:
```diff
@@ -0,0 +1,4 @@
+
+"""PR Reviewer package for automatic selection of code reviewers."""
+
+__version__ = "0.1.0"
```
I see the dynamic function below in `setup.py`. Seems like a decent amount of version maintenance code for an internal tool, for which version isn't super necessary unless you want it to be available for others to use. Even so, build numbers may suffice for internal use

tt-rkim commented on cli_tool/pr_review/review_optimizer.py:
```diff
@@ -0,0 +1,111 @@
+    # Refreshes untouched files after including reviewers
+    untouched_files = [file for file in file_owners if file['owners']]
+
+    owner_files = {}
```
Maybe rename to something like `remaining_owner_files`

Commit:
Fix reviewer selection logic to prevent duplicates

minaliuTT commented:
> 1. This could be a very useful PR
> 2. Do you have an example of what it does? Screen shot? What does select a reviewer mean?
> 3. We should avoid adding yet another top level folder to this repo. Can this tool live somewhere else?
> 
> cc: @ayerofieiev-tt @bkeith-TT

Here is the screenshort of this usage:
<img width="1023" alt="Screenshot 2025-04-03 at 9 18 21 AM" src="https://github.com/user-attachments/assets/f80d7034-bfaf-45a9-b40c-e5acb0651ee4" />

minaliuTT commented on cli_tool/README.md:
Sure, I can can own this.

ttmchiou commented:
Looks like a cool utility.
Would this be better served in its own repo?
I feel like other Tenstorrent repos can make use of it and if that's the case, they wouldn't want to include a direct dependency on tt-metal

blozano-tt reviewed (changes requested):
Thanks for trying to make the world a better place.

IMO, if this is a generic github tool, to improve developer flow, it doesn't belong in tt-metal, where it could be distracting to customers, or people trying to parse the repo.

It can be housed here: https://github.com/minaliuTT?tab=repositories

Or in some other common tools repo.

tt-rkim commented:
Yes, does this need to be in metal repo? You could maintain it separately on another repo under `tenstorrent`. If you would like help with code review, you could always pull us in, but we'd prefer to not have extra non-Metal specific tools in here, as this definitely would be useful and likely used outside Metal. @cmaryanTT We've received complaints that there's too many top-level dirs in Metal, and infra team tends to agree.

minaliuTT commented:
> Yes, does this need to be in metal repo? You could maintain it separately on another repo under `tenstorrent`. If you would like help with code review, you could always pull us in, but we'd prefer to not have extra non-Metal specific tools in here, as this definitely would be useful and likely used outside Metal. @cmaryanTT We've received complaints that there's too many top-level dirs in Metal, and infra team tends to agree.

This CLI tool is currently built specifically to support the tt-metal repository. While it could potentially be useful for other repositories down the line, we’re not quite there yet.

@bkeith-TT @cmaryanTT Let me know how your team might use this and if you think it would make more sense to move it to another repo for further development or broader use.

Commit:
Refactor setup.py, README, and review_optimizer.py

- Simplified setup.py by removing dynamic version handling from __init__.py for internal tool use.
- Updated README to clarify codeowners installation instructions.
- Renamed 'untouched_files' to 'remaining_owner_files' for better readability in review_optimizer.py.
- Cleaned up redundant POSIX-related newlines in code.

tt-rkim commented:
I don't believe it's specific to metal, though.

If you made the `repository` value a CLI argument, this could easily be moved to a different tools repo.

cmaryanTT commented:
> I don't believe it's specific to metal, though.
> 
> If you made the `repository` value a CLI argument, this could easily be moved to a different tools repo.

@tt-rkim - can you suggest a generic repo owned by TT. Do we have something for misc infra?

tt-rkim commented:
There is no common tools repo.

There are:

https://github.com/tenstorrent/tt-github-actions - for common CI tools
https://github.com/tenstorrent/tt-tools-common - for common syseng tools

I'd recommend making a new repo such as `tt-developer-tools` or something like that, that product management and developer relations can also have access to. You can open an IT ticket to do this.

tt-rkim commented:
Just discussed w/ @cmaryanTT in person. We can put this tool in the container used for IRD and people on cloud are free to pull to do their work in it, but it should still be hosted elsewhere.

cmaryanTT commented:
We should revisit this. @Aswinmcw made a GUI version https://pr-reviewer.aswinlocal.in/ and a github integrated version https://github.com/tenstorrent/tt-metal/pull/25872#issuecomment-3136736056

Aswinmcw commented:
> > We should revisit this. @Aswinmcw made a GUI version https://pr-reviewer.aswincloud.com/ and a github integrated version [#25872 (comment)](https://github.com/tenstorrent/tt-metal/pull/25872#issuecomment-3136736056) - but we need to get this running reliably on TT internal systems
> 
> I love this bot @Aswinmcw ! Why don't I see this on my PRs?

@blozano-tt the CLI comment tool is not yet landed on main, once we have a clear idea on whether we are going with GUI version in our internal network or PR comment method we can proceed with that further
